moveit_core/planning_request_adapter/src/planning_request_adapter.cpp

```diff
@@ -57,21 +57,25 @@ bool callPlannerInterfaceSolve(const planning_interface::PlannerManager& planner
     return false;
 }
 
-bool callAdapter(const PlanningRequestAdapter& adapter, const PlanningRequestAdapter::PlannerFn& planner,
-                 const planning_scene::PlanningSceneConstPtr& planning_scene,
-                 const planning_interface::MotionPlanRequest& req, planning_interface::MotionPlanResponse& res,
-                 std::vector<std::size_t>& added_path_index)
+planning_interface::MotionPlanResponse callAdapter(const PlanningRequestAdapter& adapter,
+                                                   const PlanningRequestAdapter::PlannerFn& planner,
+                                                   const planning_scene::PlanningSceneConstPtr& planning_scene,
+                                                   const planning_interface::MotionPlanRequest& req,
+                                                   std::vector<std::size_t>& added_path_index)
 {
+  planning_interface::MotionPlanResponse res;
   try
   {
-    return adapter.adaptAndPlan(planner, planning_scene, req, res, added_path_index);
+    adapter.adaptAndPlan(planner, planning_scene, req, res, added_path_index);
+    return res;
   }
   catch (std::exception& ex)
   {
     RCLCPP_ERROR(LOGGER, "Exception caught executing adapter '%s': %s\nSkipping adapter instead.",
                  adapter.getDescription().c_str(), ex.what());
     added_path_index.clear();
-    return planner(planning_scene, req, res);
+    planner(planning_scene, req, res);
```

Comment:
Are we omitting any useful bool return value here? `res` should always include success and error code, correct?

Reply:
Well, `res` always contains a `MoveItErrorCodes error_code` that indicates success or failure. https://github.com/ros-planning/moveit_msgs/blob/ros2/msg/MotionPlanResponse.msg

I don't see any useful bool from `planner(planning_scene, req, res);` -- this just updates the arguments to the `planner` function. It doesn't return anything.

Reply:
I contemplated adding something like this to check the boolean return of `planner(planning_scene, req, res);`:

```
    bool plan_result = planner(planning_scene, req, res);
    if (!plan_result)
    {
      res.error_code_.val = moveit_msgs::msg::MoveItErrorCodes::FAILURE;
    }
```

The problem I see is that this would overwrite a more specific error code with a generic FAILURE.

I guess the proper way to fix this is like:

```
  using PlannerFn =
      std::function<bool(const planning_scene::PlanningSceneConstPtr&, const planning_interface::MotionPlanRequest&,
                         planning_interface::MotionPlanResponse&)>;
```

Should become:

```
  using PlannerFn =
      std::function<void(const planning_scene::PlanningSceneConstPtr&, const planning_interface::MotionPlanRequest&,
                         planning_interface::MotionPlanResponse&)>;
```

(`MotionPlanResponse` ought to be passed by reference so it retains a value from the previous adapter.)

Reply:
Changing the return type of PlannerFn caused quite a few changes because of all the plugins that use it. We can just revert this commit if you don't think it's worthwhile. I do, though.

Reply:
Should the response be the return value of the function and not an out parameter? That would make it clear that the response replaces the bool and is where the user should look for success / failure.

Reply:
If the response is also an input (I'm sorry I haven't read the code yet, only this conversation) you could still express that by having it be both an input parameter and a return value. I doubt this function is in the hot loop of the problem and copying the `planning_interface::MotionPlanResponse` to make it more obvious to the user what is the input and output seems like an ok compromise to me.

Reply:
Although I'd prefer returning something over passing it as a mutable reference I think passing it as input + returning it is a not very intuitive. Why not merge it this way and open an issue to perform the large refactoring to return a value & remove the input parameter?

Reply:
Couldn't we simply return a `MoveItErrorCode` which [converts to bool](https://github.com/ros-planning/moveit2/blob/main/moveit_core/utils/include/moveit/utils/moveit_error_code.h#L62) and keep the short evaluations? Right now, the bool result seems to not be used at all, as an alternative I'd prefer `void` and keep an issue open.

Reply:
The tricky thing here is that `res` is used recursively [here](https://github.com/ros-planning/moveit2/blob/749c5ae0777d093fad1be6514bae2bc188fc6926/moveit_core/planning_request_adapter/src/planning_request_adapter.cpp#L143). So we NEED to pass it by mutable reference.

I'm trying to do what Henning suggests by returning a `MoveItErrorCode` in addition, but no luck so far.

Reply:
OK, I believe this is resolved now